<font size="3" color="brown">Author : UTSAV AWASTHI</font><br>
<font size="1" color="brown">About Me: http://iamawesome.com/ </font> <br>
Subscribe to my YouTube channel <a href="https://www.youtube.com/channel/UCJpp55YcP5swWj8TmIOd0OA/">here</a><br> 
For serious conversations related to career and jobs, follow me on <a href="https://www.linkedin.com/in/utsavawasthi/">LinkedIn</a><br>
For low-effort dank memes and one-liners, follow me on <a href="https://twitter.com/utsavsaidwhat">Twitter</a>

<div align="center">
    <font size = 3 color="orange" style="font-family: 'Comic Sans MS'">// ज्ञानं परमं बलम् //</font><br>
</div>
<img src = "https://cloudblogs.microsoft.com/uploads/prod/sites/32/2020/05/SQL.png" width="30%" height="30%">
<br>

<div align="center" id="contents"> <font size="5" color="orange"><b>STRUCTURE QUERY LANGUAGE</b></font> </div>
<br>

<h2> Importing Datasets </h2>
Let's import the relevant datasets on which we will run SQL queries. Press Ctrl+Enter to run the codes in each of the boxes below.<br><br>
I highly advice you to download the excel file (available <a href="https://github.com/ItsAwSM/SQL-for-Data-Analytics/blob/main/sales_database.xlsx?raw=true">here</a>) which has the schema and the data which we will query in this assignment. Spend some time to go through the schema, understand the relationships between different datasets and then analyze all the datasets in excel once. 

In [ ]:
pip install xlrd

In [1]:
import pandas as pd
import pandasql as ps

In [2]:
#Importing customer dataset
customer_dim = pd.read_excel("https://github.com/ItsAwSM/SQL-for-Data-Analytics/blob/main/sales_database.xlsx?raw=true", 
                             sheet_name="customer_dim")
customer_dim.head()

cust_id first_name last_name                     email       phone  \
0  10000001      Rahul     Gupta     rahul.gupta@gmail.com  8960722319   
1  10000002    Abhinav    Sharma  abhinav.sharma@gmail.com  8857852863   
2  10000003    Anubhav     Patel   anubhav.patel@gmail.com  8492146048   
3  10000004       Neha     Verma      neha.verma@gmail.com  8631861914   
4  10000005       John   Bernard    john.bernard@gmail.com  8215391844   

   primary_pincode  gender         dob joining_date  
0           110001    male  02-12-1977   19-02-2016  
1           400001    male  20-09-1995   01-10-2016  
2           560001    male  26-12-1986   16-08-2018  
3           600001  female  20-01-2002   28-07-2009  
4           500001    male         NaN   05-03-2017

In [3]:
#Importing product dataset
product_dim = pd.read_excel("https://github.com/ItsAwSM/SQL-for-Data-Analytics/blob/main/sales_database.xlsx?raw=true", 
                            sheet_name="product_dim")
product_dim.head()

product_id     product_name brand  category  procurement_cost_per_unit  \
0       12345     HP XYZ Mouse    HP     mouse                        650   
1       12346          HP 241H    HP    laptop                      45000   
2       12347  HP 8GB Pendrive    HP  pendrive                        350   
3       12348   Dell ABC Mouse  Dell     mouse                        350   
4       12349       Dell AX420  Dell    laptop                      30000   

     mrp  
0   1500  
1  80000  
2    800  
3   1100  
4  75000

In [4]:
#Importing delivery person dataset
delivery_person_dim = pd.read_excel("https://github.com/ItsAwSM/SQL-for-Data-Analytics/blob/main/sales_database.xlsx?raw=true",
                                    sheet_name="delivery_person_dim")
delivery_person_dim.head()

delivery_person_id             name joining_date  pincode
0             1000008  Indranoor Singh   08-03-2007   110001
1             1000002   Simon Williams   23-02-2007   400001
2             1000004    Anubhav Tyagi   16-11-2007   400001
3             1000006      Aviral Vats   16-11-2007   400001
4             1000009     Rocky Tiwari   16-11-2007   400001

In [5]:
#Importing pincode dataset
pincode_dim = pd.read_excel("https://github.com/ItsAwSM/SQL-for-Data-Analytics/blob/main/sales_database.xlsx?raw=true", 
                            sheet_name="pincode_dim")
pincode_dim.head()

pincode       city       state
0   110001  New Delhi   New Delhi
1   400001     Mumbai  Maharastra
2   560001  Bangalore   Karnataka
3   600001    Chennai  Tamil Nadu
4   500001  Hyderabad   Telangana

In [6]:
#Importing order dataset
order_dim = pd.read_excel("https://github.com/ItsAwSM/SQL-for-Data-Analytics/blob/main/sales_database.xlsx?raw=true", 
                          sheet_name="order_dim")
order_dim.head()

order_id order_type   cust_id  order_date delivery_date  tot_units  \
0  10000000001        buy  10000009  01-01-2020    03-01-2020          2   
1  10000000002        buy  10000004  01-01-2020    04-01-2020          9   
2  10000000003        buy  10000012  01-01-2020    02-01-2020          2   
3  10000000004        buy  10000012  01-01-2020    03-01-2020          8   
4  10000000005        buy  10000008  01-01-2020    06-01-2020          1   

   displayed_selling_price_per_unit  total_amount_paid  product_id  \
0                             73500             138180       12349   
1                             67500             577125       12349   
2                             60750             119070       12349   
3                               736               5824       12347   
4                             60750              57105       12349   

   delivery_person_id payment_type  delivery_pincode  
0             1000003         cash            500001  
1             1000005         cash            700001  
2             1000002   netbanking            400001  
3             1000010         cash            600001  
4             1000003        debit            500001

<b>Sample question</b><br>
You need to write the SQL query between the space mentioned in the box below and then press Ctrl+Enter to run the query. Here is an example - 

<b>Q0.</b> How many customers are male and how many are female?

In [14]:
ans = ps.sqldf("""
---------------------WRITE QUERY BELOW---------------------

SELECT 
    customers.gender,
    COUNT(customers.gender) AS Count  
FROM customer_dim AS customers

GROUP BY customers.gender


---------------------WRITE QUERY ABOVE---------------------
""")

ans

gender  Count
0  female      4
1    male     10

<h2> Questions on Basic SQL</h2>

<b>Q1.</b> How many customers do not have DOB information available?

In [35]:
#SELECT count(*) FROM us WHERE a IS NULL

ans = ps.sqldf("""
---------------------WRITE QUERY BELOW---------------------

SELECT count(*) as DOB_Missing
FROM customer_dim
WHERE dob IS NULL

---------------------WRITE QUERY ABOVE---------------------
""")

ans

DOB_Missing
0            2

<b>Q2.</b> How many customers are there in each pincode and gender combination?

In [38]:
ans = ps.sqldf("""
---------------------WRITE QUERY BELOW---------------------

SELECT 
    customers.primary_pincode as pincode,
    customers.gender,
    COUNT(*) AS Customer_Count  
FROM customer_dim AS customers

GROUP BY customers.primary_pincode,
    customers.gender

---------------------WRITE QUERY ABOVE---------------------
""")

ans

pincode  gender  Customer_Count
0   110001    male               3
1   400001    male               2
2   500001  female               1
3   500001    male               1
4   560001  female               1
5   560001    male               1
6   600001  female               1
7   600001    male               1
8   700001  female               1
9   700001    male               2

In [42]:
product_dim.head(3)

product_id     product_name brand  category  procurement_cost_per_unit  \
0       12345     HP XYZ Mouse    HP     mouse                        650   
1       12346          HP 241H    HP    laptop                      45000   
2       12347  HP 8GB Pendrive    HP  pendrive                        350   

     mrp  
0   1500  
1  80000  
2    800

<b>Q3.</b> Print product name and mrp for products which have more than 50000 MRP

In [43]:
ans = ps.sqldf("""
---------------------WRITE QUERY BELOW---------------------

SELECT
    products.product_name,
    products.mrp as MRP   
FROM product_dim as products

WHERE products.mrp > 50000

---------------------WRITE QUERY ABOVE---------------------
""")

ans

product_name    MRP
0      HP 241H  80000
1   Dell AX420  75000

<b>Q4.</b> How many delivery personal are there in each pincode?

In [ ]:
ans = ps.sqldf("""
---------------------WRITE QUERY BELOW---------------------



---------------------WRITE QUERY ABOVE---------------------
""")

ans

<b>Q5.</b> For each Pin code, print the count of orders, sum of total amount paid, average amount paid, maximum amount paid, minimum amount paid for the transactions which were paid by 'cash'. Take only 'buy' order types

In [ ]:
ans = ps.sqldf("""
---------------------WRITE QUERY BELOW---------------------



---------------------WRITE QUERY ABOVE---------------------
""")

ans

<b>Q6.</b> For each delivery_person_id, print the count of orders and total amount paid for product_id = 12350 or 12348 and total units > 8. Sort the output by total amount paid in descending order. Take only 'buy' order types

In [ ]:
ans = ps.sqldf("""
---------------------WRITE QUERY BELOW---------------------



---------------------WRITE QUERY ABOVE---------------------
""")

ans

<b>Q7.</b> Print the Full names (first name plus last name) for customers that have email on "gmail.com"?

In [ ]:
ans = ps.sqldf("""
---------------------WRITE QUERY BELOW---------------------



---------------------WRITE QUERY ABOVE---------------------
""")

ans

<b>Q8.</b> How many orders had #units between 1-3, 4-6 and 7+? Take only 'buy' order types

In [ ]:
ans = ps.sqldf("""
---------------------WRITE QUERY BELOW---------------------



---------------------WRITE QUERY ABOVE---------------------
""")

ans

<b>Q9.</b> Which pincode has average amount paid more than 150,000? Take only 'buy' order types

In [ ]:
ans = ps.sqldf("""
---------------------WRITE QUERY BELOW---------------------



---------------------WRITE QUERY ABOVE---------------------
""")

ans

<b>Q10.</b> Create following columns from order_dim data - 
<ol>
    <li>order_date</li>
    <li>Order day</li>
    <li>Order month</li>
    <li>Order year</li> 
</ol>

In [ ]:
ans = ps.sqldf("""
---------------------WRITE QUERY BELOW---------------------



---------------------WRITE QUERY ABOVE---------------------
""")

ans.sample(5)

<b>Q11.</b> How many total orders were there in each month and how many of them were returned? Add a column for return rate too.<br>
return rate = (100.0 * total return orders) / total buy orders<br>
Hint: You will need to combine SUM() with CASE WHEN

In [ ]:
ans = ps.sqldf("""
---------------------WRITE QUERY BELOW---------------------



---------------------WRITE QUERY ABOVE---------------------
""")

ans

<h2> Questions on SQL Joins</h2>

<b>Q12.</b> How many units have been sold by each brand? Also get total returned units for each brand.

In [ ]:
ans = ps.sqldf("""
---------------------WRITE QUERY BELOW---------------------



---------------------WRITE QUERY ABOVE---------------------
""")

ans

<b>Q13.</b> How many distinct customers and delivery boys are there in each state?

In [ ]:
ans = ps.sqldf("""
---------------------WRITE QUERY BELOW---------------------



---------------------WRITE QUERY ABOVE---------------------
""")

ans

<b>Q14.</b> For every customer, print how many total units were ordered, how many units were ordered from their primary_pincode and how many were ordered not from the primary_pincode. Also calulate the percentage of total units which were ordered from primary_pincode(remember to multiply the numerator by 100.0). Sort by the percentage column in descending order.

In [ ]:
ans = ps.sqldf("""
---------------------WRITE QUERY BELOW---------------------



---------------------WRITE QUERY ABOVE---------------------
""")

ans

<b>Q15.</b> For each product name, print the sum of number of units, total amount paid, total displayed selling price, total mrp of these units, and finally the net discount from selling price (i.e. 100.0 - 100.0 * total amount paid / total displayed selling price) AND the net discount from mrp (i.e. 100.0 - 100.0 * total amount paid / total mrp)

In [ ]:
ans = ps.sqldf("""
---------------------WRITE QUERY BELOW---------------------



---------------------WRITE QUERY ABOVE---------------------
""")

ans

<b>Q16.</b> For every order_id (exclude returns), get the product name and calculate the discount percentage from selling price. Sort by highest discount and print only those rows where discount percentage was above 10.10%.

In [ ]:
ans = ps.sqldf("""
---------------------WRITE QUERY BELOW---------------------



---------------------WRITE QUERY ABOVE---------------------
""")

ans

<b>Q17.</b> Using the per unit procurement cost in product_dim, find which product category has made the most profit in both absolute amount and percentage<br>
Absolute Profit = Total Amt Sold - Total Procurement Cost<br>
Percentage Profit = 100.0 * Total Amt Sold / Total Procurement Cost - 100.0

In [ ]:
ans = ps.sqldf("""
---------------------WRITE QUERY BELOW---------------------



---------------------WRITE QUERY ABOVE---------------------
""")

ans

<b>Q18.</b> For every delivery person(use their name), print the total number of order ids (exclude returns) by month in seperate columns i.e. there should be one row for each delivery_person_id and 12 columns for every month in the year

In [ ]:
ans = ps.sqldf("""
---------------------WRITE QUERY BELOW---------------------



---------------------WRITE QUERY ABOVE---------------------
""")

ans

<b>Q19.</b> For each gender - male and female - find the absolute and percentage profit (like in Q16) by product name. 

In [ ]:
ans = ps.sqldf("""
---------------------WRITE QUERY BELOW---------------------



---------------------WRITE QUERY ABOVE---------------------
""")

ans

<b>Q20.</b> Generally the more numbers of units you buy, the more discount seller will give you. For 'Dell AX420' is there a relationship between number of units ordered and average discount from selling price? Take only 'buy' order types

In [ ]:
ans = ps.sqldf("""
---------------------WRITE QUERY BELOW---------------------



---------------------WRITE QUERY ABOVE---------------------
""")

ans

And lastly, remember this :) -

<font size = 10 color="green">"</font><br>
<div align = center> 
    <font size = 5 color="green" style="font-family: 'Comic Sans MS'">Pass on what you have learned</font><br>
</div>
<div align = right> 
    <font size = 10 color="green"t>"</font><br>
</div>
<img align = "right" src = "https://orig00.deviantart.net/ff5e/f/2013/102/4/4/yoda_by_valval-d61g9rh.png" width="20%" height="20%">